File: dynamic_programming/bottom_up_optimal_binary_search_tree.py
import sys


def optimal_bst(p, q):
    results = [[sys.maxsize for _ in range(len(p) +  1)]
                for _ in range(len(p) +  1)]
    roots = [[-1 for _ in range(len(p) +  1)]
                for _ in range(len(p) +  1)]
    w = [[0 for _ in range(len(p) +  1)]
                for _ in range(len(p) +  1)]

    for i in range(1, len(p) + 1):
        results[i][i - 1] = q[i - 1]
        w[i][i - 1] = q[i - 1]

    diff = 0
    while diff < len(p):
        i = 0
        j = i + diff

        while j < len(p):
            w[i][j] = w[i][j - 1] + p[j] + q[j]
            for r in range(i, j + 1):
                if results[i][j] > results[i][r - 1] + results[r + 1][j] + w[i][j]:
                    results[i][j] = results[i][r - 1] + results[r + 1][j] + w[i][j]
                    roots[i][j] = r

            i += 1
            j += 1

        diff += 1

    return results, roots, w


if __name__ == '__main__':
    p = [0.0, 0.15, 0.10, 0.05, 0.10, 0.20]
    q = [0.05, 0.10, 0.05, 0.05, 0.05, 0.10]

    print("  i |", end='')
    for i in range(len(p)):
        print(f"{i:^5}|", end='')
    print('')
    print("p[i]|", end='')
    for _p in p:
        print(f"{_p:>5.2f}|", end='')
    print('')
    print("q[i]|", end='')
    for _q in q:
        print(f"{_q:>5.2f}|", end='')
    print('')

    results, roots, w = optimal_bst(p, q)

    diff = -1
    while diff < len(p):
        print(f"--- diff = {diff} ---")

        i = 1
        j = i + diff
        while j < len(p):
            print(f"i = {i}, j = {j} / result = {results[i][j]:.2f}, w = {w[i][j]:.2f}", end='')
            if j != i - 1:
                print(f", root = {roots[i][j]}", end='')
            print('')

            i += 1
            j += 1

        diff += 1

File: dynamic_programming/bottom_up_optimal_binary_search_tree.go
//go:build ignore

package main

import (
	"fmt"
	"math"
)

func optimalBst(p, q []float32) ([][]float32, [][]int, [][]float32) {
	results := make([][]float32, len(p)+1)
	roots := make([][]int, len(p)+1)
	w := make([][]float32, len(p)+1)

	for k := range results {
		results[k] = make([]float32, len(p)+1)
		roots[k] = make([]int, len(p)+1)
		w[k] = make([]float32, len(p)+1)
	}

	for i := 0; i < len(results); i += 1 {
		for j := 0; j < len(results); j += 1 {
			results[i][j] = float32(math.MaxInt)
		}
	}

	for i := 1; i <= len(p); i += 1 {
		results[i][i-1] = q[i-1]
		w[i][i-1] = q[i-1]
	}

	for diff := 0; diff < len(p); diff += 1 {
		i := 1
		for j := i + diff; j < len(p); j += 1 {
			w[i][j] = w[i][j-1] + p[j] + q[j]

			for r := i; r < j+1; r += 1 {
				if results[i][j] > results[i][r-1]+results[r+1][j]+w[i][j] {
					results[i][j] = results[i][r-1] + results[r+1][j] + w[i][j]
					roots[i][j] = r
				}
			}

			i += 1
		}
	}

	return results, roots, w
}

func main() {
	p := []float32{0, 0.15, 0.10, 0.05, 0.10, 0.20}
	q := []float32{0.05, 0.10, 0.05, 0.05, 0.05, 0.10}

	fmt.Print("  i |")
	for i := 0; i < len(p); i += 1 {
		fmt.Printf("%4d|", i)
	}
	fmt.Println()
	fmt.Print("p[i]|")
	for i := 0; i < len(p); i += 1 {
		fmt.Printf("%.2f|", p[i])
	}
	fmt.Println()
	fmt.Print("q[i]|")
	for i := 0; i < len(q); i += 1 {
		fmt.Printf("%.2f|", q[i])
	}
	fmt.Println()

	results, roots, w := optimalBst(p, q)
	for diff := -1; diff < len(p)-1; diff += 1 {
		fmt.Printf("--- diff = %d ---\n", diff)

		i := 1
		for j := i + diff; j < len(p); j += 1 {
			fmt.Printf("(i, j) = (%d, %d) / result = %.2f w = %.2f ", i, j, results[i][j], w[i][j])
			if j != i-1 {
				fmt.Printf("root = %d", roots[i][j])
			}
			fmt.Println()

			i += 1
		}
	}
}
